peer_gater.go

```diff
@@ -0,0 +1,302 @@
+package pubsub
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+var PeerGaterRetainStats = 6 * time.Hour
+var PeerGaterQuiet = time.Minute
+
+type peerGater struct {
+	sync.Mutex
+
+	threshold, decay   float64
+	validate, throttle float64
+
+	lastThrottle time.Time
+
+	peerStats map[peer.ID]*peerGaterStats
+	ipStats   map[string]*peerGaterStats
+
+	host host.Host
+
+	// for unit tests
+	getIP func(peer.ID) string
+}
+
+type peerGaterStats struct {
+	connected int
+	expire    time.Time
+
+	deliver, duplicate, ignore, reject float64
+}
+
+// WithPeerGater is a gossipsub router option that enables reactive validation queue
+// management.
+// The threshold parameter is the threshold of throttled/validated messages before the gating
+// kicks in.
+// The decay parameter is the (linear) decay of counters per second.
+func WithPeerGater(threshold, decay float64) Option {
```

Comment:
Should we also pass in the decay interval?

Reply:
sure, why not.